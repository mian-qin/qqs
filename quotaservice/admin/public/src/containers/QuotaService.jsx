import PropTypes from 'prop-types';
import React from 'react';
import { Component } from 'react';

import Confirmation from '../components/Confirmation.jsx';
import Namespaces from './Namespaces.jsx';
import SelectedNamespace from './SelectedNamespace.jsx';
import Sidebar from './Sidebar.jsx';

export default class QuotaService extends Component {
  fetchData() {
    const { actions, env } = this.props;
    const { fetchConfigs, fetchCapabilities } = actions;

    return fetchConfigs().then(configs => {
      if (env.capabilities) {
        return fetchCapabilities(configs.payload);
      }
      return null;
    });
  }

  componentDidMount() {
    this.fetchData();
  }

  componentWillReceiveProps(nextProps) {
    const { configs } = nextProps

    // Refetch the configs after the save because there's a chance the backend
    // may do some additional processing on the data.
    if (!configs.inRequest && !configs.error && configs.items === undefined) {
      this.fetchData();
    }
  }

  renderConfirmation() {
    const { actions, dispatch, confirm } = this.props
    return (<Confirmation
      cancel={actions.clearConfirm}
      dispatch={dispatch}
      {...confirm}
    />)
  }

  renderLoading() {
    return (
      <div className='flex-container flex-box-lg'>
        <div className='loader'>Loading...</div>
      </div>
    );
  }

  render() {
    const { selectedNamespace, configs, capabilities, confirm } = this.props
    const classNames = ['flex-container', 'fill-height-container']
    const isLoading = configs.inRequest || (capabilities && capabilities.inRequest);
    const canMakeChanges = selectedNamespace ? selectedNamespace.canMakeChanges : true;

    if (confirm) {
      classNames.push('blur')
    }

    return (
      <div>
        {confirm && this.renderConfirmation()}

        {!canMakeChanges &&
          <div className="warning">You do not have permissions to make changes to this namespace.</div>
        }

        <div className={classNames.join(' ')}>
          <Sidebar {...this.props} handleRefresh={this.fetchData.bind(this)} />
          {isLoading ? this.renderLoading() : <Namespaces {...this.props} />}
          <SelectedNamespace {...this.props} />
        </div>
      </div>
    )
  }
}

QuotaService.propTypes = {
  dispatch: PropTypes.func.isRequired,
  actions: PropTypes.object.isRequired,
  namespaces: PropTypes.object.isRequired,
  stats: PropTypes.object.isRequired,
  configs: PropTypes.object.isRequired,
  capabilities: PropTypes.object.isRequired,
  selectedNamespace: PropTypes.object,
  currentVersion: PropTypes.number.isRequired,
  env: PropTypes.object.isRequired,
  confirm: PropTypes.object
}
