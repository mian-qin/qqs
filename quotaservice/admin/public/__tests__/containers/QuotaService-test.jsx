import Promise from '../../src/promise';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import React from 'react';

import QuotaService from '../../src/containers/QuotaService.jsx';

describe('QuotaService', () => {
  let props;
  let fetchConfigsPromise;

  beforeEach(() => {
    fetchConfigsPromise = Promise.resolve({});

    props = {
      env: {
        capabilities: true,
      },
      actions: {
        fetchConfigs: () => fetchConfigsPromise,
        fetchCapabilities: jest.fn(),
      },
      dispatch: jest.fn(),
      namespaces: {},
      stats: {},
      configs: {},
      capabilities: {},
      currentVersion: 1,
    };
  });

  it('renders ok', () => {
    const component = shallow(<QuotaService {...props} />);
    const tree = toJSON(component);
    expect(tree).toMatchSnapshot();
  });

  describe('capabilities', () => {
    it('are requested', () => {
      shallow(<QuotaService {...props} />);
      return fetchConfigsPromise.then(() =>
        expect(props.actions.fetchCapabilities.mock.calls.length).toBe(1)
      );
    });

    it('are not requested', () => {
      props.env.capabilities = false;
      shallow(<QuotaService {...props} />);
      return fetchConfigsPromise.then(() =>
        expect(props.actions.fetchCapabilities.mock.calls.length).toBe(0)
      );
    });
  });
});
