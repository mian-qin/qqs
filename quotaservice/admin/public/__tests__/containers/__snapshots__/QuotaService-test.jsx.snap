// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotaService renders ok 1`] = `
<div>
  <div
    className="flex-container fill-height-container"
  >
    <Sidebar
      actions={
        Object {
          "fetchCapabilities": [Function],
          "fetchConfigs": [Function],
        }
      }
      capabilities={Object {}}
      configs={Object {}}
      currentVersion={1}
      dispatch={[Function]}
      env={
        Object {
          "capabilities": true,
        }
      }
      handleRefresh={[Function]}
      namespaces={Object {}}
      stats={Object {}}
    />
    <Namespaces
      actions={
        Object {
          "fetchCapabilities": [Function],
          "fetchConfigs": [Function],
        }
      }
      capabilities={Object {}}
      configs={Object {}}
      currentVersion={1}
      dispatch={[Function]}
      env={
        Object {
          "capabilities": true,
        }
      }
      namespaces={Object {}}
      stats={Object {}}
    />
    <SelectedNamespace
      actions={
        Object {
          "fetchCapabilities": [Function],
          "fetchConfigs": [Function],
        }
      }
      capabilities={Object {}}
      configs={Object {}}
      currentVersion={1}
      dispatch={[Function]}
      env={
        Object {
          "capabilities": true,
        }
      }
      namespaces={Object {}}
      stats={Object {}}
    />
  </div>
</div>
`;
